solr/core/src/java/org/apache/solr/core/CachingDirectoryFactory.java

```diff
@@ -421,9 +422,21 @@ public final Directory get(String path, DirContext dirContext, String rawLockTyp
         cacheValue.refCnt++;
         log.debug("Reusing cached directory: {}", cacheValue);
       }
-
-      return directory;
     }
+    return filterDirectory(directory, dirContext);
+  }
+
+  /**
+   * Potentially filters or unwraps the cached {@link Directory} depending on the intended use
+   * defined by the {@link DirContext}.
+   */
+  protected Directory filterDirectory(Directory dir, DirContext dirContext) {
```

Comment:
I created a new filterDirectory method for two reasons:
- The key for cached directories does not include the DirContext.
- This method can be overridden if a DirectoryFactory implementation needs to adapt the filtering logic.